api/txspool.go

```diff
@@ -188,190 +180,68 @@ func (a *API) getPoolTxs(c *gin.Context) {
-		Fee:     txw.Fee,
-		Nonce:   txw.Nonce,
-		// State:     txw.State,
-		Signature: txw.Signature,
-		RqAmount:  txw.RqAmount,
-		Type:      txw.Type,
-	}
-	// ToIdx
-	if txw.ToIdx != nil {
-		poolTx.ToIdx = *txw.ToIdx
-	}
-	// ToEthAddr
-	if txw.ToEthAddr == nil {
-		poolTx.ToEthAddr = common.EmptyAddr
-	} else {
-		poolTx.ToEthAddr = *txw.ToEthAddr
-	}
-	// ToBJJ
-	if txw.ToBJJ == nil {
-		poolTx.ToBJJ = common.EmptyBJJComp
-	} else {
-		poolTx.ToBJJ = *txw.ToBJJ
-	}
-	// Rq fields
-	txHasRq := false
-	// RqFromIdx
-	if txw.RqFromIdx != nil {
-		poolTx.RqFromIdx = *txw.RqFromIdx
-		txHasRq = true
-	}
-	// RqToIdx
-	if txw.RqToIdx != nil {
-		poolTx.RqToIdx = *txw.RqToIdx
-	}
-	// RqToEthAddr
-	if txw.RqToEthAddr == nil {
-		poolTx.RqToEthAddr = common.EmptyAddr
-	} else {
-		poolTx.RqToEthAddr = *txw.RqToEthAddr
-	}
-	// RqToBJJ
-	if txw.RqToBJJ == nil {
-		poolTx.RqToBJJ = common.EmptyBJJComp
-	} else {
-		poolTx.RqToBJJ = *txw.RqToBJJ
-	}
-	// RqTokenID
-	if txw.RqTokenID != nil {
-		poolTx.RqTokenID = *txw.RqTokenID
-	}
-	// RqFee
-	if txw.RqFee != nil {
-		poolTx.RqFee = *txw.RqFee
-	}
-	// RqNonce
-	if txw.RqNonce != nil {
-		poolTx.RqNonce = *txw.RqNonce
-	}
-	// RqTxID
-	if txw.RqTxID != nil {
-		poolTx.RqTxID = *txw.RqTxID
-	} else if txHasRq {
-		// If any Rq field is set RqTxID shouldn't be nil
-		return errors.New(ErrRqTxIDNotProvided)
-	}
-	// Check type, id and request id
-	_, err := common.NewPoolL2Tx(&poolTx)
+func (a *API) verifyPoolL2Tx(tx common.PoolL2Tx) error {
+	// Check type and id
+	_, err := common.NewPoolL2Tx(&tx)
 	if err != nil {
 		return tracerr.Wrap(err)
 	}
 	// Validate feeAmount
-	_, err = common.CalcFeeAmount(poolTx.Amount, poolTx.Fee)
+	_, err = common.CalcFeeAmount(tx.Amount, tx.Fee)
 	if err != nil {
 		return tracerr.Wrap(err)
 	}
 	// Get sender account information
-	account, err := a.h.GetCommonAccountAPI(poolTx.FromIdx)
+	account, err := a.h.GetCommonAccountAPI(tx.FromIdx)
 	if err != nil {
 		return tracerr.Wrap(fmt.Errorf("Error getting from account: %w", err))
 	}
 	// Validate sender:
 	// TokenID
-	if poolTx.TokenID != account.TokenID {
+	if tx.TokenID != account.TokenID {
 		return tracerr.Wrap(fmt.Errorf("tx.TokenID (%v) != account.TokenID (%v)",
-			poolTx.TokenID, account.TokenID))
+			tx.TokenID, account.TokenID))
 	}
 	// Nonce
-	if poolTx.Nonce < account.Nonce {
-		return tracerr.Wrap(fmt.Errorf("poolTx.Nonce (%v) < account.Nonce (%v)",
-			poolTx.Nonce, account.Nonce))
+	if tx.Nonce < account.Nonce {
+		return tracerr.Wrap(fmt.Errorf("tx.Nonce (%v) < account.Nonce (%v)",
+			tx.Nonce, account.Nonce))
 	}
 	// Check signature
-	if !poolTx.VerifySignature(a.cg.ChainID, account.BJJ) {
+	if !tx.VerifySignature(a.cg.ChainID, account.BJJ) {
 		return tracerr.Wrap(errors.New("wrong signature"))
 	}
 	// Check destinatary, note that transactions that are not transfers
 	// will always be valid in terms of destinatary (they use special ToIdx by protocol)
-	switch poolTx.Type {
+	switch tx.Type {
```

Comment:
One question, if`tx.Type` was not set this can be a problem. Is possible to have `tx` without `Type` set?

Reply:
Not in this part of the code: a few lines above this one we're calling `_, err := common.NewPoolL2Tx(&tx)`, which sets the `Type` if it wasn't, and returns an error if it had a wrong value